stdlib/public/runtime/ProtocolConformance.cpp

```diff
@@ -1271,5 +1280,265 @@ const Metadata *swift::findConformingSuperclass(
+          const auto& swiftProtocol = protocol.getSwiftProtocol();
+          if (!swift_conformsToProtocol(metadata, swiftProtocol)) {
+            // This type does not conform to all the specified protocol(s).
+            conforms = false;
+            break;
+          }
+        }
+      }
+      if (!conforms) {
+        continue;
+      }
+
+      body(metadata, stop, bodyContext, error);
+      if (*stop || *error) {
+        return;
+      }
+    }
+  }
+}
+
+#if defined(__MACH__)
+/// Get the conformance section of a given Mach-O image.
+///
+/// \param imageAddress The address of a Mach header or of a handle returned
+///   from \c dlopen().
+/// \param maybeDlopenHandle Whether or not \a imageAddress might be a
+///   \c dlopen() handle. If it is, an additional call to dyld is made to get
+///   its corresponding Mach header.
+///
+/// If the image does not have a conformance section, returns \c llvm::None.
+static llvm::Optional<ConformanceSection>
+_getConformanceSectionFromMachImage(const void *imageAddress,
+                                    bool maybeDlopenHandle) {
+#ifdef __LP64__
+  typedef const mach_header_64 *MachHeaderAddress;
+#else
+  typedef const mach_header *MachHeaderAddress;
+#endif
+  auto machHeader = reinterpret_cast<MachHeaderAddress>(imageAddress);
+
+  if (maybeDlopenHandle) {
+    // The input pointer might be a dlopen() handle instead of a Mach header.
+    // Try to get the corresponding Mach header.
+    auto dyldAddress = dlsym(const_cast<void *>(imageAddress), "__dso_handle");
+    if (dyldAddress) {
+      machHeader = reinterpret_cast<MachHeaderAddress>(dyldAddress);
+    }
+
+    // If dyld failed to get a Mach header from the image address, it might mean
+    // it's already a Mach header or that it could not resolve the Mac header.
+    // getsectiondata() does not check that the header passed to it is valid, so
+    // perform a basic validation before passing along the pointer we have.
+    if (machHeader->magic != MH_MAGIC && machHeader->magic != MH_MAGIC_64) {
+      return llvm::None;
+    }
+  }
+  
+  unsigned long sectionSize = 0;
+  const auto& sectionData = getsectiondata(machHeader,
+                                           SEG_TEXT,
+                                           MachOProtocolConformancesSection,
+                                           &sectionSize);
+  if (sectionData) {
+    return ConformanceSection(sectionData, sectionSize);
+  }
+
+  return llvm::None;
+}
+#endif
+
+SWIFT_CC(swift)
+static void _enumerateConformingTypesFromImage(
+  const void *imageAddress,
+  const llvm::ArrayRef<ProtocolDescriptorRef>& protocols,
+  TypeEnumerationFunction body,
+  void *bodyContext,
+  SWIFT_CONTEXT void *context,
+  SWIFT_ERROR_RESULT SwiftError **error) {
+
+  bool stop = false;
+  *error = nullptr;
+
+#if defined(__MACH__)
+  // This platform has dyld and we can use it instead of walking the section
+  // table. Takes care of the nuances of DYLD shared cache support.
```

Comment:
Linux and Windows are actually EVEN BETTER!  We have pointers to the conformance table, so we don't need to consult the loader at all.

Reply:
The tradeoff being that the dyld shared cache reduces the amount of dirty memory needed to store the conformance table. You win some, you lose some. :)